traits/ctraits.c

```diff
@@ -4017,20 +3981,10 @@ validate_trait_complex ( trait_object * trait, has_traits_object * obj,
                 }
                 PyErr_Clear();
 check_implements:
-                result = PyObject_Call( validate_implements, args, NULL );
-#if PY_MAJOR_VERSION < 3
-                rc = PyInt_AS_LONG( result );
-#else
-                rc = PyLong_AsLong( result );
+                rc = PyObject_IsInstance(value, type);
                 if( rc==-1 && PyErr_Occurred()){
-                    PyErr_Clear();
-                    Py_DECREF( args );
-                    Py_DECREF( result );
-                    break;
+                    return NULL;
```

Comment:
This is a minor behaviour change: if the `isinstance(value, type)` check raises (which should never happen in normal, non-buggy code), then the exception is now propagated. In the previous code, it would have been caught and turned into a `TraitError`.